test/diag2dwd.jl

```diff
@@ -217,6 +218,317 @@ end
+        # TODO: The `infer` type will never be here once the decapodes parsing
+        # is finished being refactored. So we don't check for it here.
+        # Note: This only returns the first type error found.
+        copies[b_idx][:type][local_name_idx] == first_type ||
+          error("The type of $(copies[b_idx][:name][local_name_idx]),
+            $(copies[b_idx][:type][local_name_idx]), in decapode
+            \"$(r[:name][b_idx])\" does not match the type of
+            $(first_box[:name][local_name_idx]), $(first_type), in decapode
+            $(r[:name][first(j_idxs)]). (Also, check that the order of the
+            decapodes you supplied matches the the order you specified in the
+            relation.)")
+      end
+    end
+
+    # Step -2: Do namespacing.
+    # Append each Var name with the name @relation gave the decapode.
+    for (copy_idx, box_name) in enumerate(r[:name])
+      for (name_idx, var_name) in enumerate(copies[copy_idx][:name])
+        copies[copy_idx][:name][name_idx] = Symbol(
+          string(box_name)*"_"*string(var_name))
+      end
+    end
+
+    # Step -1: Write over the name fields to be what was specified by
+    # @relation. (oapply cannot combine objects whose attrtypes are not equal.)
+    foreach(r[:box], r[:junction], local_ports) do b_idx, j_idx, lp_idx
+      # Get the index of the row with this name in the Var.
+      name = decapodes_vars[b_idx][2][lp_idx]
+      name = Symbol(string(r[:name][b_idx])*"_"*string(name))
+      #local_name_idx = incident(copies[b_idx], name, :name) |> only
+      local_name_idx = findfirst(==(name), copies[b_idx][:name]) |> only
+      copies[b_idx][:name][local_name_idx] = r[:variable][j_idx]
+    end
+
+    # Step 2: Start composing.
+    OpenNamedDecapodeOb, OpenNamedDecapode = OpenACSetTypes(NamedDecapode, :Var)
+    
+    OpenNamedDecapodes = map(copies, decapodes_vars) do curr_copy, d_vs
+      variables = last(d_vs)
+      FinFunctions = map(variables) do var
+        FinFunction(incident(curr_copy, var, :name), nparts(curr_copy, :Var))
+      end
+      # TODO: It would be less brittle to pass the type information from the
+      # decapodes.
+      OpenNamedDecapode{Any, Any, Symbol}(curr_copy, FinFunctions...)
+    end
+    dec = oapply(relation, OpenNamedDecapodes) |> apex
+
+    # Step 3: De-duplicate Op1s.
+    unique_by!(dec, :Op1, [:src, :tgt, :op1])
+
+    # Step 4: De-duplicate Op2s.
+    unique_by!(dec, :Op2, [:proj1, :proj2, :res, :op2])
+
+    # TODO: In case we have to de-duplicate tvars, this is the code.
+    ## Step 5: De-duplicate TVars.
+    #unique_by!(dec, :TVar, [:incl])
+
+    # TODO: There must be a final step where you remove namespacing.
+
+    return dec
+  end
+
+  # TODO: Is this the best way to get multiple dispatch to work?
+  function compose_decapodes(decapode_vars, relation::RelationDiagram)
+    compose_decapodes([decapode_vars], relation)
+  end
+
+  # This is the example from the "Overview" page in the docs.
```

Comment:
These tests don't rely on the other diag2dwd tests, so they should be in their own file.